docs/design-docs/03-pipeline-merge.md

```diff
@@ -0,0 +1,405 @@
+      configPath: setters-updated.yaml // upstream overrides local 
+```
+
+Similarly, the upstream version wins if both upstream and local are updated, else local is preserved.
+
+```yaml
+Original:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+
+Updated upstream:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1.2
+      configPath: setters.yaml
+
+Local:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1.1
+      configPath: setters.yaml
+
+Expected Output:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1.2
+      configPath: setters-updated.yaml
+```
+
+This might not be what all users expect. But this is the default behavior in case 
+of conflict while merging normal resources as well. So we can inherit the behavior 
+and be consistent. In order to provide more visibility to the users, we can add 
+log messages in cases of such conflicts and intimate users about the updated value. 
+In the future, we can add support to a different conflict strategy of `--local-wins`
+as an option to the kpt pkg update command.
+
+#### More examples with expected output
+
+Newly added upstream functions are appended at the end.
+
+```yaml
+Original:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+
+Updated upstream:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+    - image: gcr.io/kpt-fn/generate-folders:v0.1
+
+Local:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+    - image: gcr.io/kpt-fn/set-namespace:v0.1
+      configMap:
+        namespace: foo
+
+Expected output:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+    - image: gcr.io/kpt-fn/set-namespace:v0.1
+      configMap:
+        namespace: foo
+    - image: gcr.io/kpt-fn/generate-folders:v0.1
+```
+
+If a function is deleted upstream and not changed on the local, it will be deleted on local.
+
+```yaml
+Original:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+    - image: gcr.io/kpt-fn/generate-folders:v0.1
+
+
+Updated upstream:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+
+Local:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+    - image: gcr.io/kpt-fn/generate-folders:v0.1
+    - image: gcr.io/kpt-fn/set-namespace:v0.1
+      configMap:
+        namespace: foo
+
+Expected output:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+    - image: gcr.io/kpt-fn/set-namespace:v0.1
+      configMap:
+        namespace: foo
+```
+
+Same function declared multiple times: If the same function is declared multiple 
```

Comment:
How do handle situations where functions might have been reordered either in local or upstream? I'm worried that using a merge key that is the combination of image and input type might make it hard to understand/predict the behavior in some cases. For example, what happens if we don't have a unique merge key in a list?

Reply:
Updated the doc. Best way is to use `name` field.